README.md
# Sample Size Estimation
``` r
install.packages("sse")
library(sse)


psi <- powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
              n = seq(from = 20, to = 60, by = 10))
powFun <- function(psi) {
  return(power.t.test(n = n(psi)/2, 
                      delta = theta(psi),
                      sig.level = 0.05)$power)
}
calc <- powCalc(psi, statistic = powFun)
pow <- powEx(x = calc, theta = 1, power = 0.9)
plot(pow)
```


ChangeLog
version 0.7-15 2019/0
	change refine() to only refine the theta and xi of the example (so
 	far it was only the theta) and adding a test to
	inst/slowTests/testthat/test-advanced.R
	
	change update() to correctly evaluating n.iter also after changes
	in vector of n, theta or xi and adding several tests to
	inst/slowTests/testthat/test-updateResample.R
version 0.7-14 2019/09/23
	adding error message in update() if n.iter is smaller than
          available (also adding corresponding test)
	DESCRIPTION: removing 'lintr' from Suggests
version 0.7-13 2019/08/22
	vignette: removing unused options
	man pages: several changes to harmonize wording and improve structure
version 0.7-12 2019/08/13
	splitting of tests into tests at tests/ and inst/slowTests to
	  reduce the time used for R CMD check
	correcting some spelling errors in man pages
	correcting some spelling in warning/error messages
	changing the sequence of asserting that arguments are fine for
	  (argument 'smooth' in plot() and of 'lm.range' in powEx())
	some resolving of lintr issues (line length etc.)
version 0.7-11 2019/07/19
	removing tests of parallel computing since these fail on some
	  systems and since parallel computing is only under development
version 0.7-10 2019/07/01
        update(): corrections that allow updating different aspects of n, theta
	and xi in combination
	refine(): Change argument name 'n.iter' to 'factor' (to show that there
	  is a difference to n.iter in calc).
	refine(): Do not change slot 'iter' of the power-object.
	Handling of forceDivisor, a best guess for the divisor is added based
	  on the chosen n-values.
	Refining many test.
	Adding some test for the current implementation of parallel computing.


pkg/sse/tests/testthat.R
library(testthat)
library(sse)


test_check("sse")


## some conventions:

## if expect_error is used, separat lines allow to easily read and check the error massage manually:
## e.g.:
## expect_error(
##     powPar(n = seq(from = 20, to = 60, by = 2))  ## READ ERROR MESSAGE
## )


tests/testthat/test-standard.R
context("Standard use of sse")

library(testthat)
library(sse)
## as used in the vignette(examples)
psi <- powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
              n = seq(from = 20, to = 60, by = 2))
psi2 <- powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
              n = seq(from = 20, to = 60, by = 2),
              xi = c(0.05, 0.1))
##
powFun.power <- function(psi){
  return(power.t.test(n = n(psi) / 2,
                      delta = theta(psi),
                      sig.level = 0.05)$power)
}
powFun.power2 <- function(psi){
  return(power.t.test(n = n(psi) / 2,
                      delta = theta(psi),
                      sig.level = xi(psi2))$power)
}
powFun.resample <- function(psi){
  x <- rnorm(n(psi) / 2)
  y <- rnorm(n(psi) / 2) + theta(psi)
  return(wilcox.test(x = x, y = y)$p.value < 0.05)
}


##
calc.power <- powCalc(psi, statistic = powFun.power)
calc.power2 <- powCalc(psi2, statistic = powFun.power2)
calc.resample <- powCalc(psi, statistic = powFun.resample, n.iter = 33)
# calc.resample <- powCalc(psi, statistic = powFun.resample, n.iter = -1) ## FIXME catch error by power.R

#
calc.resample2 <- update(calc.resample, theta = 1, n.iter = 99)


##
pow.power <- powEx(calc.power, theta = 1, power = 0.9)
pow.power2 <- powEx(calc.power2, theta = 1, xi = 0.05, power = 0.9)
pow.resample <- powEx(calc.resample, theta = 1, power = 0.9)
pow.resample <- powEx(pow.resample, theta = 0.8, power = 0.9)
pow.resample2 <- powEx(calc.resample2, theta = 1, power = 0.9)
##
inspect(pow.resample)
inspect(pow.resample2)
##

plot(pow.resample)

pow.refined <- refine(pow.resample)


inspect(pow.refined)
plot(pow.refined)
#pow.refined@iter <- as.integer(33)
update(pow.refined, n = seq(from = 30, to = 60, by = 5))

pow.refined.newEx <- powEx(pow.refined, theta = 1, power = 0.9)

pow.refined2 <- refine(pow.resample, factor = 99)
inspect(pow.refined2)
plot(pow.refined2)
### --------------------------------- TEST PLOTS  -->>  CHECK THEM VISUALLY
##



plot(pow.power,
     xlab = "Effect size",
     ylab = "Total sample size",
     )

## default is without smooting ?
plot(pow.resample,
     xlab = "Effect size",
     ylab = "Total sample size")

## default is with smooting
plot(pow.resample,
     xlab = "Effect size",
     ylab = "Total sample size",
     smooth = TRUE)

## sequence of "at" elements
## is the right line emph (power = 0.9)?
plot(pow.power,
     xlab = "Effect size",
     ylab = "Total sample size",
     at = seq(from = 0.8, to = 0.95, by = 0.05))

## is the right line emph (power = 0.8)?
plot(pow.power,
     xlab = "Effect size",
     ylab = "Total sample size",
     at = seq(from = 0.8, to = 0.95, by = 0.05),
     example = FALSE)

## xlim, correct range and overall look of labels etc?
plot(pow.power,
     xlab = "Effect size",
     ylab = "Total sample size",
     xlim = c(0.79, 1.21))

## ylim, correct range and overall look of labels etc?
plot(pow.power,
     xlab = "Effect size",
     ylab = "Total sample size",
     ylim = c(30, 50))

## xlim and ylim, correct range and overall look of labels etc?
plot(pow.power,
     xlab = "Effect size",
     ylab = "Total sample size",
     xlim = c(0.79, 1.21),
     ylim = c(30, 50))

## default method for resampling is "lm" -> check: is method lm is used?
inspect(pow.resample)
inspect(pow.resample2)


## method = "step" is possible for resampling -> check: is method step is used?
inspect(powEx(calc.resample, theta = 1, power = 0.9, method = "step"))
inspect(powEx(calc.resample2, theta = 1, power = 0.9, method = "step"))

inspect(powEx(calc.resample2,
              theta = 1,
              power = 0.9,
              method = "step",
              forceDivisor = 2))
inspect(
    powEx(calc.resample2,
          theta = 1,
          power = 0.9,
          method = "step",
          forceDivisor = TRUE)
)




### --------------------------------- TESTS
test_that("powPar", {
  # and not an object of class powCalc
  expect_error( # no theta
      powPar(n = seq(from = 20, to = 60, by = 2))  ## READ ERROR MESSAGE
  )
  expect_error( # an n that can not be coerced to an integer
      powPar(n  = calc.power, theta = seq(from = 0.5, to = 1.5, by = 0.05))
      ## READ ERROR MESSAGE
  )
  ## n as integer is dangerous but allowed
  ## expect_error(
  ##    powPar(n = c(TRUE, FALSE), theta = seq(from = 0.5, to = 1.5, by = 0.05))
  ## )
  expect_error(
      powPar(used.as.theta = seq(from = 0.5, to = 1.5, by = 0.05),
             n = seq(from = 20, to = 60, by = 2))
  )
  expect_error(
      powPar(used.as.theta = seq(from = 0.5, to = 1.5, by = 0.05),
             n = seq(from = 20, to = 60, by = 2),
             theta.name = "as.theta")
  )
  expect_error(
      powPar(n = c(10, 15, 15, 20),
             theta = seq(from = 0.5, to = 1.5, by = 0.05))
  )
  expect_error(
      powPar(n = c(10, 15, 20),
             theta = c(0.5, 0.6, 0.6))
  )
  expect_error(
      powPar(n = c(10, 15, 20),
             theta = c(0.5, 0.6),
             xi = c(1, 1, 2))
  )
  
})



test_that("powCalc", {
  ## expects an object of class powPar and not an object of class powCalc
  expect_error(
      powCalc(seq(from = 0, to = 100, by = 10), powFun1)
  )
  ## n.iter of length > 1
  expect_warning(
      powCalc(psi, statistic = powFun.resample, n.iter = c(4, 33))
  )
  ## ## default is using cluster
  ## expect_message(
  ##     calc.resample <- powCalc(psi,
  ##               statistic = powFun.resample, n.iter = 3, cluster = TRUE)
  ##    , "using cluster"
  ## )
  ## ## if cluster = FALSE
  ## expect_message(
  ##     calc.resample <- powCalc(psi,
  ##              statistic = powFun.resample, n.iter = 3, cluster = FALSE)
  ##     , "not using cluster"
  ## )
})



test_that("powEx", {

  ## power out of observed range
  pow.r1 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.1, to = 0.5, by = 0.1),
                 n = seq(from = 20, to = 60, by = 2)),
          statistic = powFun.power),
      theta = 0.1, power = 0.9)
  expect_error(
      plot(pow.r1)  
  )

  ## power is a constant
  pow.c1 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.5, to = 1.5, by = 0.1),
                 n = seq(from = 20, to = 60, by = 2)),
          statistic = function(psi){return(0.9)}),
      theta = 1, power = 0.9)
  expect_error(
      plot(pow.c1)  
  )

  ## a wrong method, e.g. linear, is not allowed
  expect_error(
      powEx(calc.power, theta = 1, method = "linear")  ## READ ERROR MESSAGE
  )

  ## theta is not in the range used
  expect_error(
      powEx(calc.power, theta = 2)  ## READ ERROR MESSAGE
  )
  ## 
  expect_error(
      powEx(calc.power, theta = 2, forceDivisor = -2)  ## READ ERROR MESSAGE
  )
  expect_error(
      powEx(calc.power, theta = 2, endpoint = "ep2")  ## READ ERROR MESSAGE
  )
  ## calc object does not use xi but example for xi provided
  expect_warning(
      powEx(calc.power, theta = 1, xi = 7)  ## READ WARNING MESSAGE
  )

  ## from the internal usage of construct.pwEx
  expect_warning(
      powEx(calc.power, theta = 1, endpoint = c("ep1", "ep2"))
  )
  expect_error(
      powEx(calc.power, theta = 1, power = 1.1)
  )
  expect_warning(
      powEx(calc.power, theta = 1, power = c(0.8, 0.9))
  )
  expect_warning(
      powEx(calc.power, theta = 1, drop = c(0.1, 0.15))
  )
  expect_warning(
      powEx(calc.power, theta = c(1, 2))
  )
  expect_warning(
      powEx(calc.power2, theta = 1, xi = c(0.05, 0.1))
  )
  expect_warning(
      powEx(calc.power, theta = 1, method = "step", lm.range = 0.2)
  )
  expect_error(
      powEx(calc.power, theta = 1, method = "lm", lm.range = 1.2)
  )
    expect_warning(
      powEx(calc.power, theta = 1, lm.range = c(0.2, 0.3))
  )
    expect_warning(
        powEx(calc.power, theta = 1, forceDivisor = 2.000001)
    )
    expect_warning(
        powEx(calc.power, theta = 1, forceDivisor = 2.000001, method = "lm")
    )
    expect_error(# FIXME
        powEx(calc.power, theta = 1, forceDivisor = list(1:2))
    )

  expect_error(
      powEx(calc.power, theta = 1, drop = 1.5)
    )

  
})


test_that("refine", {

  ## only if
  expect_error(
      refine(pow.power),
      strwrap(
          "Additional iterations for the chosen example are only
            meaningful if the object was created using resampling.",
          prefix = " ", initial = ""))
  ## does not work for objects of class calc
  expect_error(
      refine(calc.resample, factor = 0.5)  ## READ WARNING MESSAGE
  )
  ## factor must be larger than 1
  expect_error(
      refine(pow.resample, factor = 0.5)  ## READ WARNING MESSAGE
  )
  ## factor must be larger than 1
  expect_error(
      refine(pow.resample, factor = 1)  ## READ WARNING MESSAGE
  )
})


### ------------------------------------------------------------------ plot
test_that("plot", {

  ##
  expect_warning(
      plot(pow.power, ylim = c(30, 40))  ## READ WARNING MESSAGE
  )

  ## auto detect best divisor
  pow.d0 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
                 n = c(13, 26)),
          statistic = powFun.power),
      theta = 1, power = 0.5, forceDivisor = TRUE)
  expect_equal(pow.d0@divisor, as.integer(13))


  ## divisior changes number return
  pow.d1 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
                 n = seq(from = 20, to = 60, by = 1)),
          statistic = powFun.power),
      theta = 1, power = 0.7, forceDivisor = 4)#, method = "lm")
  expect_message(
      plot(pow.d1)
  )
  expect_message(
      inspect(pow.d1)
  )
  ## divisior changes number return
  pow.d1 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
                 n = seq(from = 20, to = 60, by = 1)),
          statistic = powFun.power),
      theta = 1, power = 0.7, forceDivisor = 4, method = "lm")
  expect_message(
      plot(pow.d1)
  )
  expect_message(
      inspect(pow.d1)
  )

  ## with length of theta == 1, plotting is not possible
  pow.t1 <- powEx(
      powCalc(
          powPar(theta = 0.5,
                 n = seq(from = 20, to = 60, by = 2)),
          statistic = powFun.power),
      theta = 0.5, power = 0.9)
  expect_error(
  plot(pow.t1)
  )

  ## with length of n == 1, plotting is not possible
  pow.n1 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.5, to = 1.5, by = 0.05),
                 n = 20),
          statistic = powFun.power),
      theta = 0.5, power = 0.9)
  expect_error(
      plot(pow.n1)
  )
  
  ## example outside of plotting range
  pow.o1 <- powEx(calc.power, theta = 0.5, power = 0.9)
  expect_warning(
      plot(pow.o1)
  )
  expect_warning(
      inspect(pow.o1)
  )

  ## argument smooth
  expect_warning(
      plot(pow.power, smooth = c(1, 2))
    )
  expect_error(
      plot(pow.power, smooth = -1)
  )

  ## argument at
  expect_error(
      plot(pow.power, at = FALSE)
  )
  expect_error(
      plot(pow.power, at = -0.5)
  )

  expect_warning(
      plot(pow.power, xlim = c(0.1, 0.8))
  )
  
})


## a helper function for latex strings
## prep.tex <- function(string){
##   new.string <- gsub("\\", "", string, fixed = TRUE)
##   new.string <- gsub("{", "", new.string, fixed = TRUE)
##   new.string <- gsub("}", "", new.string, fixed = TRUE)
##   return(new.string)
## }


### ------------------------------------------------------------------ tex
test_that("tex methods", {
  expect_match(tex(pow.power, "drop"), "0~\\\\%", fixed = TRUE)
  expect_match(tex(pow.power, "sampling"),
               "$n_{i=1,...,21} = 20, ..., 60$", fixed = TRUE)
  expect_equal(tex(pow.power, "nRec"), 46)
  expect_equal(tex(pow.power, "nEval"), 46)
  expect_equal(tex(pow.power, "theta"), 1)
  expect_equal(tex(pow.power, "xi"), as.numeric(NA))
  expect_equal(tex(pow.power, "power"), 0.9)
  expect_equal(tex(pow.power, "n.iter"), as.numeric(NA))
  expect_error(tex(pow.power, "hallo"))
})



### ------------------------------------------------------------------ show
test_that("show", {
  show(pow.power)                                           # READ CAREFULLY

  ## information about resampling
  show(pow.resample)                                           # READ CAREFULLY
  
  ## with additional parameters in powPar
  pow.s1 <- powEx(
      powCalc(
          powPar(theta = seq(from = 0.1, to = 0.5, by = 0.1),
                 n = seq(from = 20, to = 60, by = 2),
                 extra = 1:8),
          statistic = powFun.power),
      theta = 0.1, power = 0.9)

  show(pow.s1)
})




### ---------------------------------------------------------------- powFunGen
test_that("powFunGen", {
  
powFun.matrix <- function(psi){
  return(matrix(1:16))
}
expect_error(
    calc.power <- powCalc(psi, statistic = powFun.matrix)
)



})


pkg/sse/tests/testthat/test-lintr.R
## context("lintr")



## # "lintr_trailing_semicolon_linter" fails in bad1
## # "lintr_attach_detach_linter"  fails in bad2
## # "lintr_setwd_linter"  fails in bad2
## # "lintr_sapply_linter"  fails in bad2
## # "lintr_library_require_linter" fails in bad2
## # "lintr_seq_linter" fails in bad3
## # not failing yet:
## # "lintr_assignment_linter"
## # "lintr_line_length_linter" 

## gp_lintrs <- c("lintr_assignment_linter", "lintr_line_length_linter", 
##                "lintr_trailing_semicolon_linter", 
##                "lintr_attach_detach_linter", "lintr_setwd_linter", 
##                "lintr_sapply_linter", "lintr_library_require_linter", 
##                "lintr_seq_linter")

## bad1 <- system.file("bad1", package = "goodpractice")
## gp_bad1 <- gp(bad1, checks = gp_lintrs)
## res_bad1 <- results(gp_bad1)

## gp_bad2 <- gp("bad2", checks = gp_lintrs)
## res_bad2 <- results(gp_bad2)

## gp_bad3 <- gp("bad3", checks = gp_lintrs)
## res_bad3 <- results(gp_bad3)

## get_result <- function(res, check) res$result[res$check == check]



## test_that("lintr_assignment_linter", {
  
##   expect_true(get_result(res_bad1, "lintr_assignment_linter"))
##   # TODO expectation/example where the check fails
  
## })


## library(lintr)
## test_that("code quality and style", {
## #  skip_if_not_installed("lintr")
## #  skip_on_travis()
## #  skip_on_cran()
##   lintr::expect_lint_free(linters = with_defaults(
##                               line_length_linter(120)
##                           ))
## })


## context("lints")

## ## #if (requireNamespace("lintr", quietly = TRUE)) {


## ## library(lintr)

## ## # enforce camelCase rather than snake_case
## ## with_defaults(camel_case_linter = NULL,
## ##               snake_case_linter)
## ##                                         # change the default line length cutoff
## ## with_defaults(line_length_linter = line_length_linter(120))

## ## lint("power.R")
## ## , linters =  list(line_length_linter(80)))


## test_that("Package Style", {
## ##     lintr::expect_lint_free(#linters = list(
## ##                             #    e = lintr::line_length_linter(100))
## ##                             )
## ##   })
## ## #}
## lintr::lint_package("sse", 
##              linters = with_defaults(
##                  ## absolute_paths_linter = NULL, # checks that no absolute paths are used.
##                  ## assignment_linter = NULL, #checks that '<-' is always used for  assignment.
##                  ## closed_curly_linter = NULL, #check that closed curly braces should always be on their own line unless they follow an else.
##                  ## commas_linter = NULL, #check that all commas are followed byspaces, but do not have spaces before them.
##                  commented_code_linter = NULL, #checks that there is no commented code outside roxygen blocks
##                  ## infix_spaces_linter = NULL, #check that all infix operators have spaces around them.
##    #              line_length_linter(120), #check the line length of both comments and code is less than length.
##                  ## no_tab_linter = NULL, #check that only spaces are used, never tabs.
##                  ## object_usage_linter = NULL, #checks that closures have the proper usage using ‘checkUsage’.  Note this runs ‘eval’ on the code, so do not use with untrusted code.
##                  camel_case_linter = NULL, #check that objects are not in camelCase.
##                  snake_case_linter = NULL, #check that objects are not in snake_case.
##                  multiple_dots_linter = NULL, #check that objects do not have multiple.dots.
##    #              object_length_linter = NULL, #check that objects do are not very long.not have.multiple.dots.
##                  ## open_curly_linter = NULL, #check that opening curly braces are never on their own line and are always followed by a newline.
##                  ## single_quotes_linter = NULL, #checks that only single quotes are used to delimit string contestants.
##                  ## spaces_inside_linter = NULL, #check that parentheses and square brackets do not have spaces directly inside them.
##                  ## spaces_left_parentheses_linter = NULL, #check that all left parentheses have a space before them unless they are in a function call.
##                  ## trailing_blank_lines_linter = NULL, #check there are no trailing blank lines.
##                  ## trailing_whitespace_linter = NULL, #check there are no trailing whitespace characters.
##                  object_camel_case_linter = NULL
##   )
## )
## })


pkg/sse/R/AllGenerics.R
## calculates the array with the power, stores it into powCalc
setGeneric(name = "powCalc",
           def = function(object, statistic, n.iter = NA, cluster = FALSE){
             standardGeneric("powCalc")
           })

## for generating the example object in one step (replacing powEx and
## merge that are kept for back compatibility (and internal usage))
setGeneric(name = "powEx",
           def = function(x, theta,
                          xi = NA,
                          endpoint = NA,
                          power = 0.9,
                          drop = 0,
                          method = c("default", "lm", "step"),
                          lm.range = NA,
                          forceDivisor = FALSE){
             standardGeneric("powEx")
           })

setGeneric(name = "powFunGen",
           def = function(object, ...){
             standardGeneric("powFunGen")
           })

setGeneric(
           name = "workhorse",
           def = function(object, ...){
             standardGeneric("workhorse")
           })

## for programming on powPar parameters
setGeneric(name = "pp",
           def = function(x, name){
             standardGeneric("pp")
           })

## for programming on powPar parameters
setGeneric(name = "n",
           def = function(x){
             standardGeneric("n")
           }
           #,
           # useAsDefault = function(object = pow, ...){object@n.act}
           )

setGeneric(name = "refine",
           def = function(object, n.iter = 10){
             standardGeneric("refine")
           })

## for programming on powPar parameters
setGeneric(name = "theta",
           def = function(x){
             standardGeneric("theta")
           })

## for programming on powPar parameters
setGeneric(name = "xi",
           def = function(x){
             standardGeneric("xi")
           })

## estimats n from objects of class "power"
setGeneric(name = "sampleSize",
           def = function(x, ...) {
             standardGeneric("sampleSize")
           })
## estimats n from power
setGeneric(name = "exDat",
           def = function(x, y, ...) {
             standardGeneric("exDat")
           })

## estimats n from power
setGeneric(name = "inspect",
           def = function(object) {
             standardGeneric("inspect")
           })

## for reporting the sampling schema
setGeneric(name = "tex",
           def = function(x, type, ...){
             standardGeneric("tex")
           })


pkg/sse/inst/slowTests/testthat/test-updateResample.R
context("Update, resampling")

library(sse)
library(testthat)
psi1 <- powPar(theta = seq(from = 0, to = 1, by = 0.1),
               n = seq(from = 0, to = 100, by = 10))
##
powFun2 <- function(psi) {
  n  <-  n(psi)
  theta  <-  theta(psi)
  return(as.logical( (n * theta) %% 2))
}
powFun3 <- function(psi) {
  n  <-  n(psi)
  theta  <-  theta(psi)
  return(as.logical(n %% 2))
}

## correct result for

result1.2.x <- array(round(seq(from = 0, to = 1, by = 0.1) %*%
                           t(seq(from = 0, to = 100, by = 10)) %% 2, 10),
                     dim = c(11, 11, 1, 1))
result1.2.n2 <- array(as.logical(seq(from = 0, to = 1, by = 0.1) %*%
                                 t(seq(from = 50, to = 100, by = 5)) %% 2),
                      dim = c(11, 11, 1, 1))
## result1.2.n5 <- array(as.logical(seq(from = 0, to = 1, by = 0.1) %*%
##                                  t(seq(from = 50, to = 150, by = 5)) %% 2),
##                       dim = c(21, 11, 1, 1))
##                                         # FIXME
result1.3.x <- array(0, dim = c(11, 11, 1, 1))

##
calc1.2.1 <- powCalc(psi1, statistic = powFun2, n.iter = 99)

calc1.3.1 <- powCalc(psi1, statistic = powFun3, n.iter = 99)

calc1.2.1.nc <- powCalc(psi1, statistic = powFun2, n.iter = 99, cluster = FALSE)
                                        # as calc1.2.1 but without cluster
##
n.new <- as.integer(seq(from = 0, to = 100, by = 5))
n.new2 <- as.integer(seq(from = 50, to = 100, by = 5))
n.new5 <- as.integer(seq(from = 50, to = 150, by = 5))
theta.new <- seq(from = 0, to = 1, by = 0.05)
theta.new4 <- 0.51

##
pow1.2.1 <- powEx(calc1.2.1, theta = 1, power = 0.9)

pow1.2.1.rf <- refine(pow1.2.1)

test_that("update without any changes", {
#
  calc1.2.2 <- update(calc1.2.1)
  calc1.3.2 <- update(calc1.3.1)
  expect_equal(calc1.2.2@core, result1.2.x, check.attributes = FALSE)
  expect_equal(calc1.3.2@core, result1.3.x, check.attributes = FALSE)
})

test_that("n.iter", {
  ## increasing
  calc1.2.3 <- update(calc1.2.1, n.iter = 500)
  expect_equal(calc1.2.3@core, result1.2.x, check.attributes = FALSE)
  expect_equal(calc1.2.3@iter, 500)
  ## n.iter length > 1
  expect_warning(
      update(calc1.2.1, n.iter = c(500,100))
  )
## update of n.iter with value < than available
  expect_error(
      update(calc1.2.1, n.iter = 50)
  )
})


test_that("increasing n within existing range of n", {
#
  calc1.2.3 <- update(calc1.2.1, n = n.new)
})


test_that("increasing theta within existing range of theta", {
#
  calc1.2.4 <- update(calc1.2.1, theta = theta.new)
  #expect_equal(calc1.1.4@core, result1.1.t, check.attributes = FALSE)
})



test_that("increasing theta and n within existing range of theta and n", {
#
  calc1.2.5 <- update(calc1.2.1, n.iter = NA, n = n.new, theta = theta.new)
#  expect_equal(calc1.1.5@core, result1.1.nt, check.attributes = FALSE)
})



test_that("update statistic only", {
#
  calc1.2.6 <- update(calc1.2.1, statistic = powFun2)
  expect_equal(calc1.2.6@core, result1.2.x, check.attributes = FALSE)
})



test_that("update statistic and increasing n within existing range of n", {
#
  calc1.2.7 <- update(calc1.2.1, statistic = powFun2, n = n.new)
  calc1.2.7 <- update(calc1.2.7, n = n.new)
#  expect_equal(calc1.2.7@core, result1.2.n, check.attributes = FALSE)
})



test_that("updating n within smaller range of n", {
#
  calc1.2.8 <- update(calc1.2.1, n = n.new2)
 # expect_equal(array(as.logical(calc1.2.8@core), dim = c(11,11,1,1)),
 # result1.2.n2, check.attributes = FALSE)
})

test_that("updating n with new range of n only partially matching", {
#
  calc1.2.8 <- update(calc1.2.1, n = n.new5)
 # expect_equal(array(as.logical(calc1.2.8@core), dim = c(11,11,1,1)),
 # result1.2.n2, check.attributes = FALSE)
})

test_that("not using cluster", {
#
  expect_equal(calc1.2.1.nc@core, result1.2.x, check.attributes = FALSE)
  expect_equal(calc1.2.1.nc@core, result1.2.x, check.attributes = FALSE)
})


test_that("new theta of length 1 and increasing n.iter", {
#
  calc1.2.9 <- update(calc1.2.1.nc, theta = theta.new4, n.iter = 99)
  calc1.2.10 <- update(calc1.2.9, theta = theta.new4, n.iter = 100)
})

test_that("updating n with new range of n only partially matching", {
#
  calc1.2.11 <- update(calc1.2.1, n = n.new5)
  #expect_equal(array(as.logical(calc1.2.11@core), dim = c(21,11,1,1)),
  # result1.2.n5, check.attributes = FALSE)
})


pkg/sse/inst/slowTests/testthat/test-advanced.R
context("Advanced applications")

library(sse)
library(testthat)
### ------------------------------------------------------------------
### with pilot data and several endpoints
pilot.data <- rnorm(1000)
#
psi <- powPar(F.hat = pilot.data,
              delta = seq(from = 0.5, to = 1.5, by = 0.05),
              n = seq(from = 20, to = 50, by = 2),
              theta.name = "delta")

powFun <- function(psi){
   a <- sample(pp(psi, "F.hat"), size = n(psi) / 2, replace = TRUE)
   b <- sample(pp(psi, "F.hat"), size = n(psi) / 2, replace = TRUE) + theta(psi)
   w <- wilcox.test(a, b)$p.value < 0.05
   t <- t.test(a, b)$p.value < 0.05
   return(c(w = w, t = t))
   }

calc <- powCalc(psi, statistic = powFun, n.iter = 99)

pow.w <- powEx(calc, theta = 1, drop = 0.1, endpoint = "w")

plot(pow.w, smooth = 0.5,
     xlab = expression(paste("Delta, ", delta)),
     ylab = "Total sample size",
     main = "Wilcoxon Test")

pow.t <- powEx(calc, theta = 1, drop = 0.1, endpoint = "t")

plot(pow.t, smooth = 0.5,
     xlab = expression(paste("Delta, ", delta)),
     ylab = "Total sample size",
     main = "T- Test",
     ylim = c(30, 40))

## parametric resampling:
psi.parametric <- powPar(delta = seq(from = 0.5, to = 1.5, by = 0.05),
                         xi = seq(from = 0.5, to = 1.5, by = 0.5),
                         n = seq(from = 20, to = 50, by = 2),
                         theta.name = "delta")


powFun.parametric <- function(psi){
  a <- rnorm(n(psi), mean = 0, sd = xi(psi))
  b <- rnorm(n(psi), mean = theta(psi), sd = xi(psi))
  w <- wilcox.test(a, b)$p.value < 0.05
  t <- t.test(a, b)$p.value < 0.05
   return(c(w = w, t = t))
}

calc.parametric <- powCalc(psi.parametric,
                           statistic = powFun.parametric,
                           n.iter = 99)

pow.t.parametric <- powEx(calc.parametric,
                          xi = 1,
                          theta = 1,
                          drop = 0.1,
                          endpoint = "t")

pow.t.parametric.xi05 <- powEx(calc.parametric,
                               xi = 0.5,
                               theta = 1,
                               drop = 0.1,
                               endpoint = "t")

plot(pow.t.parametric)

show(pow.t.parametric)

### --------------------------------- TESTS
test_that("endpoints", {
#
  ## selecting an endpoint that does not exist
  expect_error(
      powEx(calc, theta = 1, drop = 0.1, endpoint = "T")  ## READ ERROR MESSAGE
  )
  expect_equal(pow.t.parametric@endpoint.example, "t")
})


test_that("powPar with data", {
#
  ## data extracted from psi is like the original
  expect_equal(pp(psi, "F.hat"), pilot.data)
})


test_that("refine", {
  pow.w.rf <- refine(pow.w)
  expect_equal(pow.w.rf@iter, pow.w@iter)
  expect_equal(pow.w.rf@iter.example, pow.w@iter * 10)
})

test_that("powEx", {
  ## choosing and endpoint for example that is not part of the calc-object
  expect_error(powEx(calc.power, theta = 2, endpoint = "s"))  ## READ MESSAGE
})

test_that("plot", {
  ## power is constant at 1
  expect_error(
      plot(pow.t.parametric.xi05)
  )
})

test_that("List as return", {
  powFun.list <- function(psi){
    a <- sample(pp(psi, "F.hat"), size = n(psi) / 2, replace = TRUE)
    b <- sample(pp(psi, "F.hat"),
                size = n(psi) / 2, replace = TRUE) + theta(psi)
    w <- wilcox.test(a, b)$p.value < 0.05
    t <- t.test(a, b)$p.value < 0.05
    length(c(a,b) %% 2)
    return(list(power = c(w = w, t = t), size = sum(c(a,b) %/% 2)))
}
expect_error(    
    calc.power <- powCalc(psi, statistic = powFun.list)
)
})
